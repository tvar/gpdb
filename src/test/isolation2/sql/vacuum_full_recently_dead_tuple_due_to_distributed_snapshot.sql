-- Test to validate bug fix for vacuum full and distributed snapshot.
--
-- Scenarios is where locally on segment tuple's xmin and xmax is lower than
-- OldestXmin and hence safely can be declared DEAD but based on distributed
-- snapshot HeapTupleSatisfiesVacuum() returns HEAPTUPLE_RECENTLY_DEAD. Later
-- though while moving the tuples around as part of vacuum, distributed snapshot
-- was not consulted but instead xmin was only checked against OldestXmin
-- raising the "ERROR: updated tuple is already HEAP_MOVED_OFF".
create table test_recently_dead_utility(a int, b int, c text);
-- Insert enough data that it doesn't all fit in one page.
insert into test_recently_dead_utility select 1, g, 'foobar' from generate_series(1, 1000) g;
-- Perform updates to form update chains and deleted tuples.
update test_recently_dead_utility set b = 1;
update test_recently_dead_utility set b = 1;

-- Run VACUUM FULL in utility mode. It sees some of the old, updated, tuple
-- versions as HEAPTUPLE_RECENTLY_DEAD, even though they are safely dead because
-- localXidSatisfiesAnyDistributedSnapshot() is conservative and returns 'true'
-- in utility mode. Helps to validate doesn't surprise the chain-following logic
-- in VACUUM FULL.
2U: vacuum full verbose test_recently_dead_utility;
2U: select count(*) from test_recently_dead_utility;
2U: set gp_select_invisible=1;
-- print to make sure deleted tuples were not cleaned due to distributed
-- snapshot to make test is future proof, if logic in
-- localXidSatisfiesAnyDistributedSnapshot() changes for utility mode.
2U: select count(*) from test_recently_dead_utility;
2U: set gp_select_invisible=0;
