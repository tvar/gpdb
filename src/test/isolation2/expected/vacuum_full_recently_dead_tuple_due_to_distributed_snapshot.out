-- Test to validate bug fix for vacuum full and distributed snapshot.
--
-- Scenarios is where locally on segment tuple's xmin and xmax is lower than
-- OldestXmin and hence safely can be declared DEAD but based on distributed
-- snapshot HeapTupleSatisfiesVacuum() returns HEAPTUPLE_RECENTLY_DEAD. Later
-- though while moving the tuples around as part of vacuum, distributed snapshot
-- was not consulted but instead xmin was only checked against OldestXmin
-- raising the "ERROR: updated tuple is already HEAP_MOVED_OFF".
create table test_recently_dead_utility(a int, b int, c text);
CREATE
-- Insert enough data that it doesn't all fit in one page.
insert into test_recently_dead_utility select 1, g, 'foobar' from generate_series(1, 1000) g;
INSERT 1000
-- Perform updates to form update chains and deleted tuples.
update test_recently_dead_utility set b = 1;
UPDATE 1000
update test_recently_dead_utility set b = 1;
UPDATE 1000

-- Run VACUUM FULL in utility mode. It sees some of the old, updated, tuple
-- versions as HEAPTUPLE_RECENTLY_DEAD, even though they are safely dead because
-- localXidSatisfiesAnyDistributedSnapshot() is conservative and returns 'true'
-- in utility mode. Helps to validate doesn't surprise the chain-following logic
-- in VACUUM FULL.
2U: vacuum full verbose test_recently_dead_utility;
VACUUM
2U: select count(*) from test_recently_dead_utility;
count
-----
1000 
(1 row)
2U: set gp_select_invisible=1;
SET
-- print to make sure deleted tuples were not cleaned due to distributed
-- snapshot to make test is future proof, if logic in
-- localXidSatisfiesAnyDistributedSnapshot() changes for utility mode.
2U: select count(*) from test_recently_dead_utility;
count
-----
2000 
(1 row)
2U: set gp_select_invisible=0;
SET
